icu4c/source/tools/gencnval/gencnval.c

```diff
@@ -273,6 +273,11 @@ main(int argc, char* argv[]) {
 
     const char* sourcedir = options[SOURCEDIR].value;
     if (sourcedir != NULL && *sourcedir != 0) {
+        if (strlen(sourcedir) + strlen(path) + 1 >= 512) {
+            fprintf(stderr,
+                    "Length of sourcedir + path is too long, must be <= 510.\n");
```

Comment:
I don't like magic numbers because updates can get these values out of sync. Please switch `>= 512` to `> sizeof(pathBuf)` so that the relationship is clear. You also already have the +1 for the null.

Please also provide a more informative error message that a person unfamiliar with the tool can understand. For example, this might be more informative:

```
 fprintf(stderr, "The source filename is too long. It must be less than %i in bytes.\n", sizeof(pathBuf) - 1);
```

Optionally, you can consider switching the size of pathBuf to MAX_PATH/PATH_MAX from limits.h, but that may take time to get right from all of the platforms. A length smaller than necessary is fine, since most of the ICU tools should not be used normally.

Reply:
Markus's suggestion is also an acceptable path forward.

Reply:
> I don't like magic numbers because updates can get these values out of sync. Please switch `>= 512` to `> sizeof(pathBuf)` so that the relationship is clear. You also already have the +1 for the null.

On top of the null byte at the end, a byte for the file separation character might be needed:

```
        if(*(end-1)!=U_FILE_SEP_CHAR) {
            *(end++)=U_FILE_SEP_CHAR;
        }
```

right?

Reply:
> > I don't like magic numbers because updates can get these values out of sync. Please switch `>= 512` to `> sizeof(pathBuf)` so that the relationship is clear. You also already have the +1 for the null.
> 
> On top of the null byte at the end, a byte for the file separation character might be needed:
> 
> ```
>         if(*(end-1)!=U_FILE_SEP_CHAR) {
>             *(end++)=U_FILE_SEP_CHAR;
>         }
> ```
> 
> right?

So this should work, I think:

```c
    if (sourcedir != NULL && *sourcedir != 0) {
        if (strlen(sourcedir) + strlen(path) + 2 > sizeof(pathbuf)) {
            fprintf(stderr,
                    "The source file name is too long, it must be less than %d in bytes.\n",
                    sizeof(pathbuf) - 1);
            exit(U_ILLEGAL_ARGUMENT_ERROR);
        }
        char *end;
        uprv_strcpy(pathBuf, sourcedir);
        end = uprv_strchr(pathBuf, 0);
        if(*(end-1)!=U_FILE_SEP_CHAR) {
            *(end++)=U_FILE_SEP_CHAR;
        }
        uprv_strcpy(end, path);
        path = pathBuf;
    }
```

Reply:
Yes. Using sizeof addresses George's well-founded dislike of magic numbers.
Doesn't matter whether you test `+ 1 >= sizeof` or `+ 2 > sizeof`.

In the error message, you will need to cast the value to (int) so that it fits the %d.

We can switch this code to icu::CharString in a follow-up PR.

Reply:
Yes the proposed fix in the comment works for me too.

Reply:
> Yes the proposed fix in the comment works for me too.

I pushed again with the fix as in the comment (adding the cast to int and fixing the typos, pathbuf -> pathBuf).